TheFunctionChangedMyMind/README-en.md

```diff
@@ -0,0 +1,322 @@
+
+One specific case is when you want to start a new `Activity` on your app. This means –in short– launching a whole new screen or an external application. To do this, you first create an `Intent` based on your `Context` and pass the `Activity` class you want to start:
+
+```kotlin
+val myIntent = Intent(context, MyNextActivity::class.java)
+```
+
+Then you can start `MyNextActivity` from another `Activity` simply using this intent: `startActivity(myIntent)`
+
+If your application relays on activities for showing different sections/views of your application then probably you will end with a lot of these calls distributed around. In some different variants, but all of them with the same structure:
+
+```kotlin
+val someIntent = Intent(context, SomeActivity::class.java)
+//Do stuff with the intent
+startActivity(someIntent)
+
+//And if you don't do things with the intent and you are in an activity
+startActivity(
+    Intent(this, MyNextActivity::class.java)
+)
+```
+
+## My first problem
+
+To be honest, there is no problem with this code. Not at all. But I have seen this happening in a lot of applications, you start having these calls distributed around without control, then you suddenly have 2 or 3 places doing the same, setting the same intent with a different variable name but same parameters and then doing the same call. To solve these, most people usually do one of two things:
+
+1. Wrap it around a single method for each case
+2. Put all under a single class/object and just pass needed parameters for repeated cases
+
+This means, for case 1:
+
+```kotlin
+/** Inside your class that will have the calls **/
+fun startNext() {
+    startActivity(
+        Intent(this, MyNextActivity::class.java)
+    )
+}
+
+fun startSome() {
+    val someIntent = Intent(this, SomeActivity::class.java)
+    //Do things with intent
+    startActivity(someIntent)
+}
+```
+
+And for case 2:
+
+```kotlin
+object AppRouter {
+    fun startNext(context: Context) {
+        startActivity(
+            Intent(context, MyNextActivity::class.java)
+        )
+    }
+
+    fun startSome(context: Context) {
+        val someIntent = Intent(context, SomeActivity::class.java)
+        //Do things with intent
+        startActivity(someIntent)
+    }
+}
+```
+
+You can add different layers of complexity but this is the base for most solutions: **have a single point for launching/starting new activities**. In my case, I always ended having case 1 all around. Having functions here and there seemed like a good solution, until I ended with 120 functions doing almost the same all around, which is not bad, but seeing that `::class.java` even inside of different methods really bothered me. Plus, seeing 3 or 4 functions inside the same class with almost the same body also feels wrong.
```

Comment:
You can add several layers of complexity, but this is the base for most solutions:

In my case, I always ended **up** with 120...